Kube support for gMSA

saikiranakula-amzn opened the pull request:
Added logic to parse kubeconfig

*Issue #, if available:*

*Description of changes:*
Parsing the kube config file to create the data get service account information

*Testing*
[ec2-user@ip-x-x-x-x build]$ sudo ./credentials-fetcherd --self_test
run diagnostic set
krb_files_dir = /var/credentials-fetcher/krbdir
logging_dir = /var/credentials-fetcher/logging
unix_socket_dir = /var/credentials-fetcher/socket
read meta data file test is successful
<2>krb file path contains invalid characters
read invalid metadata test is successful
<2>directory path for kerberos tickets is not provided
krb dir not found test is successful
write meta data info to file test is successful
parsing kube config test successful


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

smhmhmd commented on auth/kerberos/src/kube2krb.py:
```diff
@@ -0,0 +1,20 @@
+json_string=json.dumps(json_dict)
+yaml_out = yaml.dump(json_dict)
+f = open(sys.argv[1], "w")
+f.write(yaml_out)
```
Add kube apply to secret yaml and folder
Take kube folder as one more input and do a kube apply on the folder

saikiranakula-amzn committed:
Added renewal kubeflow

saikiranakula-amzn committed:
Parse kube config

Added logic to parse kubeconfig

saikiranakula-amzn commented on config/src/config.cpp:
```diff
@@ -60,6 +60,15 @@ int parse_options( int argc, const char* argv[], creds_fetcher::Daemon& cf_daemo
+        {
+            cf_daemon.kube_config_file_path = vm["kube_config_file_path"].as<std::string>();
+            std::cout
+                << "Option kubeconfig domainlees config file is selected"
```
have a logger to journalctl

smhmhmd commented on api/tests/kubeconfig.json:
```diff
@@ -0,0 +1,40 @@
+      "ServiceAccountName": "webapp03",
+      "path_to_cred_spec_json": "contoso_webapp03.json",
+      "domainless_user": "",
+      "krb_ticket_location": "",
```
Change variable to say that user should not add a value

smhmhmd committed:
Fix compilation issue for kube self-test

	[ec2-user@EC2AMAZ-N9W1Aw build]$ sudo ./credentials-fetcherd --self_test
	run diagnostic set
	krb_files_dir = /var/credentials-fetcher/krbdir
	logging_dir = /var/credentials-fetcher/logging
	unix_socket_dir = /var/credentials-fetcher/socket
	read meta data file test is successful
	<2>krb file path contains invalid characters
	read invalid metadata test is successful
	<2>directory path for kerberos tickets is not provided
	krb dir not found test is successful
	write meta data info to file test is successful
	webapp01
	/var/credentials-fetcher/krbdir/07d9d29b64b5fbc4d970/webapp01
	webapp03
	/var/credentials-fetcher/krbdir/07d9d29b64b5fbc4d970/webapp03
	parsing kube config test successful

saikiranakula-amzn committed:
Added new metadata cache for kube

saikiranakula-amzn committed:
Added pod with secret yaml modification

saikiranakula-amzn committed:
commit to support EKS DIY on credentials-fetcher

smhmhmd committed:
Add documentation on EKS integration

Description is current for this PR.

saikiranakula-amzn committed:
Command line arg option kube apply

smhmhmd committed:
Update EKS.md

Add step-8 and minor updates to other steps.

smhmhmd committed:
Fix rpmbuild and add kubeconfig file to rpm

* Added kubeconfig.json as credentials_fetcher_kubeconfig.json
* Location is /etc/credentials_fetcher_kubeconfig.json
* Needs code changes to fix references to kube config

Test:
[ec2-user@EC2AMAZ-3UYjZe ~]$ rpm -qlp /home/ec2-user/rpmbuild/RPMS/x86_64/credentials-fetcher-1.2.0-1.amzn2023.x86_64.rpm
/etc/credentials_fetcher_kubeconfig.json
/usr/lib/.build-id
/usr/lib/.build-id/3a
/usr/lib/.build-id/3a/de8e58809dccdb157f26a4fd2545156b5419b3
/usr/lib/systemd/system/credentials-fetcher.service
/usr/sbin/credentials-fetcherd
/usr/sbin/credentials_fetcher_utf16_private.exe
/usr/sbin/credentials_fetcher_utf16_private.runtimeconfig.json
/usr/share/doc/credentials-fetcher
/usr/share/doc/credentials-fetcher/CONTRIBUTING.md
/usr/share/doc/credentials-fetcher/NOTICE
/usr/share/doc/credentials-fetcher/README.md
/usr/share/licenses/credentials-fetcher
/usr/share/licenses/credentials-fetcher/LICENSE
[ec2-user@EC2AMAZ-3UYjZe ~]$

saikiranakula-amzn committed:
test

saikiranakula-amzn committed:
Fix renewal logic

saikiranakula-amzn committed:
Add kube support

smhmhmd committed:
Remove install step in spec file

The %files section in spec file copies kube config.json

[ec2-user@EC2AMAZ-3UYjZe credentials-fetcher]$ rpm -qlp  /home/ec2-user/rpmbuild/RPMS/x86_64/credentials-fetcher-1.3.0-1.amzn2023.x86_64.rpm
/etc/credentials_fetcher_kubeconfig.json
/usr/lib/.build-id
/usr/lib/.build-id/ee
/usr/lib/.build-id/ee/15e825e3518485c79f4f513a79aa8c5cdbf72a
/usr/lib/systemd/system/credentials-fetcher.service
/usr/sbin/credentials-fetcherd
/usr/sbin/credentials_fetcher_utf16_private.exe
/usr/sbin/credentials_fetcher_utf16_private.runtimeconfig.json
/usr/share/doc/credentials-fetcher
/usr/share/doc/credentials-fetcher/CONTRIBUTING.md
/usr/share/doc/credentials-fetcher/NOTICE
/usr/share/doc/credentials-fetcher/README.md
/usr/share/licenses/credentials-fetcher
/usr/share/licenses/credentials-fetcher/LICENSE

smhmhmd committed:
Add kube python file to rpm

The kube python file does the 'kube apply' to the secrets
and optionally to the pods as well.

smhmhmd committed:
Fix python script that converts krb to kube secrets

* Add syslog logging
* Use kube_config option to fix error

saikiranakula-amzn committed:
Added support to apply pods

* Apply pods if the defintion is provided
* Fix to handle ticket creation failures

smhmhmd committed:
Add config file

* /etc/credentials_fetcher.conf has global configuration regarding modes
* Added conf file to rpm package with default of ecs mode
* Conf file has comments on the modes

Testing:
* Built RPM, but not tested
* The next change in this PR to test

saikiranakula-amzn committed:
Fix meta file format

* File format fixes kube mode

saikiranakula-amzn committed:
Meta json file for kube yaml mappings

smhmhmd committed:
Read configuration from config file

This affects EKS specifically, without a config file, the mode defaults
to the previous ECS mode.

Testing:
I tested with the following conf file and EKS config
/etc/credentials_fetcher.conf

 credentials_fetcher_mode = "domain_joined_and_eks_mode"
 credentials_fetcher_kubeconfig_path = "/etc/credentials_fetcher_kubeconfig.json"

$ sudo cat /etc/credentials_fetcher_kubeconfig.json
{
  "ServiceAccountMappings": [
    {
      "ServiceAccountName": "webapp01",
      "path_to_cred_spec_json": "/home/ec2-user/customertest_webapp01.json",
      "domainless_user": "",
      "kube_context": [
        {
          "kube_context_name": "name",
          "path_to_kube_pod_yaml_file": "/home/ec2-user/pod_with_secret.yaml",
          "path_to_kube_secret_yaml_file": "/home/ec2-user/secret1.yaml"
        }
      ]
    }
  ]
}

$ sudo cat /home/ec2-user/customertest_webapp01.json
{
  "CmsPlugins":  [
    "ActiveDirectory"
  ],
  "DomainJoinConfig":  {
    "Sid":  "S-1-5-21-4217655605-3681839426-3493040985",
    "MachineAccountName":  "webapp01",
    "Guid":  "af602f85-d754-4eea-9fa8-fd76810485f1",
    "DnsTreeName":  "customertest.local",
    "DnsName":  "customertest.local",
    "NetBiosName":  "customertest"
  },
  "ActiveDirectoryConfig":  {
    "GroupManagedServiceAccounts":  [
      {
        "Name":  "webapp01",
        "Scope":  "customertest.local"
      },
      {
        "Name":  "webapp01",
        "Scope":  "customertest"
      }
    ]
  }
}

smhmhmd committed:
Update EKS doc

Add brief steps and notes

smhmhmd committed:
Add gcc flag -fstack-protector-all

smhmhmd committed:
Add note in EKS.md

saikiranakula-amzn committed:
change lease folder for eks mode

smhmhmd committed:
Simplified configuration and reformatted files

* /etc/credentials_fetcher.conf has these configurable options:
credentials_fetcher_mode, credentials_fetcher_krbdir_path and credentials_fetcher_krbfile_suffix
See credentils_fetcher.conf for description

* Added credentials_fetcher_krbfile_suffix for usecases like krb5cc_1000

* Tested grpc mode using gmsa_test_client by setting credentials_fetcher_mode = "domain_joined_and_grpc_mode"

* Tested eks config file mode by setting credentials_fetcher_mode = "domain_joined_and_eks_mode"
  and the kubeconfig json as before

smhmhmd committed:
Added credentials_fetcher_fixed_lease_name_dir option to conf file

gRPC configuration uses lease directories
This option can be used in the config file mode to have fixed paths
such as /var/credentials-fetcher/krbdir/my_kerberos_tickets_dir

saikiranakula-amzn committed:
Add change to take config_file mode

* Tested EKS mode

smhmhmd committed:
Tweaks to text in /etc/credentials_fetcher.conf

No code change

smhmhmd committed:
Add code-coverage collection

Code coverage requires rebuild

Build credentials-fetcher using:
* mkdir build && cd build && cmake -DCOVERAGE:STRING=1 ../ && make

Get code-coverage tools using following steps:
* Download lcov from https://github.com/linux-test-project/lcov
* cd lcov && sudo make install
* Install perl modules
** curl -L https://cpanmin.us | perl - --sudo App::cpanminus.
** sudo perl -MCPAN -e 'install "DateTime"'
** sudo perl -MCPAN -e 'install "Devel::Cover"'
** sudo perl -MCPAN -e 'install "Digest::MD5"'
** sudo perl -MCPAN -e 'install "File::Spec"'
** sudo perl -MCPAN -e 'install "JSON::XS"'
** sudo perl -MCPAN -e 'install "Memory::Process"'
** sudo perl -MCPAN -e 'install "Time::HiRes"'

The perl modules take a while to install.

Steps to collect code-coverage:
* Run credentials-fetcher as "sudo credentials-fetcherd" from build
  directory
* Run the sample program to exercise daemon
** Make sure to change domain string in gmsa_test_client.cpp and rebuild:
   "sudo ./build/api/tests/gmsa_test_client"
* Send kill signal to dump code coverage data
** ps -elf | grep cred
** kill -SIGUSR1 2464771 2464773
* Copy gcda and gcno files:
**  find . -name "*gcno*" -exec cp {} . \;*
**  find . -name "*gcda*" -exec cp {} . \;*
* Run lcov
** lcov --capture --directory . --output-file coverage.info
* Get html output
** genhtml coverage.info --output-directory out

smhmhmd committed:
Fix conditional compile for code-coverage

Part of previous commit
